File: problems/tier2-intermediate/016-amazon-locker/solution.js
'use strict';

/* Amazon Locker — allocation/deposit/retrieval with expiry + notifications. */

class LockerOp {
  constructor(kind, s1 = '', s2 = '', i1 = 0, i2 = 0) {
    this.kind = kind;
    this.s1 = s1;
    this.s2 = s2;
    this.i1 = i1;
    this.i2 = i2;
  }
}

// Locker sizes — keep ordered SMALL < MEDIUM < LARGE
const SMALL = 'SMALL';
const MEDIUM = 'MEDIUM';
const LARGE = 'LARGE';

class Locker {
  constructor(lockerId, size) {
    this.locker_id = lockerId;
    this.size = size;
    this.occupied = false;
  }
}

class DepositRecord {
  constructor(lockerId, packageId, pickupCode, depositTime) {
    this.locker_id = lockerId;
    this.package_id = packageId;
    this.pickup_code = pickupCode;
    this.deposit_time = depositTime;
  }
}

class SmallestFitStrategy {
  allocate(packageSize, available) {
    let tryOrder;
    if (packageSize === SMALL) tryOrder = [SMALL, MEDIUM, LARGE];
    else if (packageSize === MEDIUM) tryOrder = [MEDIUM, LARGE];
    else tryOrder = [LARGE];
    for (const sz of tryOrder) {
      const q = available.get(sz);
      if (q && q.length > 0) {
        return q.shift();
      }
    }
    return '';
  }
}

class CapturingChannel {
  constructor(log) {
    this.log = log;
  }
  notify(packageId, message) {
    this.log.push(packageId + ': ' + message);
  }
}

class LockerSystem {
  constructor() {
    this.lockers = new Map();
    this.available = new Map([
      [SMALL, []],
      [MEDIUM, []],
      [LARGE, []],
    ]);
    this.active_deposits = new Map();
    this.strategy = new SmallestFitStrategy();
    this.channels = [];
    this.code_counter = 0;
    this.expiry_hours = 0;
  }

  _generate_code() {
    this.code_counter += 1;
    return 'CODE-' + String(this.code_counter);
  }

  _notify_all(packageId, message) {
    for (const ch of this.channels) ch.notify(packageId, message);
  }

  _free_locker(lockerId) {
    if (this.lockers.has(lockerId)) {
      const locker = this.lockers.get(lockerId);
      locker.occupied = false;
      this.available.get(locker.size).push(lockerId);
    }
  }

  add_locker(lockerId, size) {
    this.lockers.set(lockerId, new Locker(lockerId, size));
    this.available.get(size).push(lockerId);
  }

  deposit_package(packageId, size, depositTime = 0) {
    const lockerId = this.strategy.allocate(size, this.available);
    if (!lockerId) return '';
    this.lockers.get(lockerId).occupied = true;
    const code = this._generate_code();
    this.active_deposits.set(code, new DepositRecord(lockerId, packageId, code, depositTime));
    this._notify_all(packageId, 'Package ' + packageId + ' deposited. Code: ' + code);
    return code;
  }

  retrieve_package(code) {
    if (!this.active_deposits.has(code)) return false;
    const rec = this.active_deposits.get(code);
    this._free_locker(rec.locker_id);
    this.active_deposits.delete(code);
    return true;
  }

  set_code_expiry(hours) {
    this.expiry_hours = hours;
  }

  check_expired(currentTime) {
    const expired = [];
    if (this.expiry_hours <= 0) return expired;
    for (const code of [...this.active_deposits.keys()]) {
      const rec = this.active_deposits.get(code);
      if (currentTime - rec.deposit_time > this.expiry_hours * 3600) {
        this._free_locker(rec.locker_id);
        expired.push(rec.package_id);
        this._notify_all(rec.package_id, 'Package ' + rec.package_id + ' expired. Locker freed.');
        this.active_deposits.delete(code);
      }
    }
    return expired;
  }

  add_notification_channel(channel) {
    this.channels.push(channel);
  }
}

function _lsize_from(s) {
  if (s === 'S') return SMALL;
  if (s === 'M') return MEDIUM;
  return LARGE;
}

function locker_simulate(ops) {
  const out = [];
  let sys = new LockerSystem();
  let codes = new Array(32).fill('');
  let chanLog = [];
  let chan = null;
  let lastExpired = [];
  for (const op of ops) {
    const k = op.kind;
    if (k === 'new') {
      sys = new LockerSystem();
      codes = new Array(32).fill('');
      chanLog = [];
      chan = null;
      lastExpired = [];
      out.push('ok');
    } else if (k === 'add_locker') {
      sys.add_locker(op.s1, _lsize_from(op.s2));
      out.push('ok');
    } else if (k === 'deposit') {
      const code = sys.deposit_package(op.s1, _lsize_from(op.s2), op.i1);
      if (op.i2 >= 0 && op.i2 < codes.length) codes[op.i2] = code;
      out.push(code);
    } else if (k === 'code_at') {
      out.push(codes[op.i2]);
    } else if (k === 'retrieve') {
      out.push(sys.retrieve_package(codes[op.i2]) ? 'ok' : 'fail');
    } else if (k === 'retrieve_id') {
      out.push(sys.retrieve_package(op.s1) ? 'ok' : 'fail');
    } else if (k === 'set_expiry') {
      sys.set_code_expiry(op.i1);
      out.push('ok');
    } else if (k === 'check_expired') {
      lastExpired = sys.check_expired(op.i1);
      out.push(String(lastExpired.length));
    } else if (k === 'expired_at') {
      if (op.i2 >= 0 && op.i2 < lastExpired.length) out.push(lastExpired[op.i2]);
      else out.push('');
    } else if (k === 'add_chan') {
      chan = new CapturingChannel(chanLog);
      sys.add_notification_channel(chan);
      out.push('ok');
    } else if (k === 'chan_log_size') {
      out.push(String(chanLog.length));
    } else if (k === 'chan_log_contains') {
      const found = chanLog.some(entry => entry.includes(op.s1));
      out.push(found ? 'yes' : 'no');
    } else {
      out.push('unknown:' + k);
    }
  }
  return out;
}

module.exports = { LockerOp, locker_simulate };

File: problems/tier2-intermediate/016-amazon-locker/solution.py
"""Amazon Locker — allocation/deposit/retrieval with expiry + notifications."""

from collections import deque


class LockerOp:
    def __init__(self, kind, s1="", s2="", i1=0, i2=0):
        self.kind = kind
        self.s1 = s1
        self.s2 = s2
        self.i1 = i1
        self.i2 = i2


# Locker sizes — keep ordered SMALL < MEDIUM < LARGE
SMALL = "SMALL"
MEDIUM = "MEDIUM"
LARGE = "LARGE"


class Locker:
    def __init__(self, locker_id, size):
        self.locker_id = locker_id
        self.size = size
        self.occupied = False


class DepositRecord:
    def __init__(self, locker_id, package_id, pickup_code, deposit_time):
        self.locker_id = locker_id
        self.package_id = package_id
        self.pickup_code = pickup_code
        self.deposit_time = deposit_time


class SmallestFitStrategy:
    def allocate(self, package_size, available):
        if package_size == SMALL:
            try_order = [SMALL, MEDIUM, LARGE]
        elif package_size == MEDIUM:
            try_order = [MEDIUM, LARGE]
        else:
            try_order = [LARGE]
        for sz in try_order:
            q = available.get(sz)
            if q:
                return q.popleft()
        return ""


class CapturingChannel:
    def __init__(self, log):
        self.log = log

    def notify(self, package_id, message):
        self.log.append(package_id + ": " + message)


class LockerSystem:
    def __init__(self):
        self.lockers = {}
        self.available = {SMALL: deque(), MEDIUM: deque(), LARGE: deque()}
        self.active_deposits = {}
        self.strategy = SmallestFitStrategy()
        self.channels = []
        self.code_counter = 0
        self.expiry_hours = 0

    def _generate_code(self):
        self.code_counter += 1
        return "CODE-" + str(self.code_counter)

    def _notify_all(self, package_id, message):
        for ch in self.channels:
            ch.notify(package_id, message)

    def _free_locker(self, locker_id):
        if locker_id in self.lockers:
            locker = self.lockers[locker_id]
            locker.occupied = False
            self.available[locker.size].append(locker_id)

    def add_locker(self, locker_id, size):
        self.lockers[locker_id] = Locker(locker_id, size)
        self.available[size].append(locker_id)

    def deposit_package(self, package_id, size, deposit_time=0):
        locker_id = self.strategy.allocate(size, self.available)
        if not locker_id:
            return ""
        self.lockers[locker_id].occupied = True
        code = self._generate_code()
        self.active_deposits[code] = DepositRecord(locker_id, package_id, code, deposit_time)
        self._notify_all(package_id, "Package " + package_id + " deposited. Code: " + code)
        return code

    def retrieve_package(self, code):
        if code not in self.active_deposits:
            return False
        rec = self.active_deposits[code]
        self._free_locker(rec.locker_id)
        del self.active_deposits[code]
        return True

    def set_code_expiry(self, hours):
        self.expiry_hours = hours

    def check_expired(self, current_time):
        expired = []
        if self.expiry_hours <= 0:
            return expired
        for code in list(self.active_deposits.keys()):
            rec = self.active_deposits[code]
            if current_time - rec.deposit_time > self.expiry_hours * 3600:
                self._free_locker(rec.locker_id)
                expired.append(rec.package_id)
                self._notify_all(rec.package_id, "Package " + rec.package_id + " expired. Locker freed.")
                del self.active_deposits[code]
        return expired

    def add_notification_channel(self, channel):
        self.channels.append(channel)


def _lsize_from(s):
    if s == "S":
        return SMALL
    if s == "M":
        return MEDIUM
    return LARGE


def locker_simulate(ops):
    out = []
    sys = LockerSystem()
    codes = [""] * 32
    chan_log = []
    chan = None
    last_expired = []
    for op in ops:
        k = op.kind
        if k == "new":
            sys = LockerSystem()
            codes = [""] * 32
            chan_log = []
            chan = None
            last_expired = []
            out.append("ok")
        elif k == "add_locker":
            sys.add_locker(op.s1, _lsize_from(op.s2))
            out.append("ok")
        elif k == "deposit":
            code = sys.deposit_package(op.s1, _lsize_from(op.s2), op.i1)
            if 0 <= op.i2 < len(codes):
                codes[op.i2] = code
            out.append(code)
        elif k == "code_at":
            out.append(codes[op.i2])
        elif k == "retrieve":
            out.append("ok" if sys.retrieve_package(codes[op.i2]) else "fail")
        elif k == "retrieve_id":
            out.append("ok" if sys.retrieve_package(op.s1) else "fail")
        elif k == "set_expiry":
            sys.set_code_expiry(op.i1)
            out.append("ok")
        elif k == "check_expired":
            last_expired = sys.check_expired(op.i1)
            out.append(str(len(last_expired)))
        elif k == "expired_at":
            if 0 <= op.i2 < len(last_expired):
                out.append(last_expired[op.i2])
            else:
                out.append("")
        elif k == "add_chan":
            chan = CapturingChannel(chan_log)
            sys.add_notification_channel(chan)
            out.append("ok")
        elif k == "chan_log_size":
            out.append(str(len(chan_log)))
        elif k == "chan_log_contains":
            found = any(op.s1 in entry for entry in chan_log)
            out.append("yes" if found else "no")
        else:
            out.append("unknown:" + k)
    return out
